File: arrays/in-place-modification/t.js
/**
 * remove duplicates from sorted array
 */
// using extra space
function remove_duplicates_from_sorted_array_in_place_and_return_length(arr){
    let k=1
    for(let i=1;i<arr.length;i++){
        if(arr[i]!==arr[i-1]){
            arr[k]=arr[i]
            k++
        }
    }
    console.log(k);
}
// remove_duplicates_from_sorted_array_in_place_and_return_length([1,1,2,2,3,4,4,4,5,5,6])

function remove_occurence_of_element_and_return_length(arr,val){
    let k=0
    for(let i=0;i<arr.length;i++){
        if(arr[i]!==val){
            arr[k]=arr[i]
            k++
        }
    }
    console.log(k);
}
// remove_occurence_of_element_and_return_length([1,2,3,2,2,2,1,2],2)
function rotate_array_in_place(nums,k){
    k=k%nums.length
   let i=0,j=nums.length-1
    while(i<j){
        // [nums[i],nums[j]]=[nums[j],nums[i]]
        let t=nums[i]
        nums[i]=nums[j]
        nums[j]=t
        i++,j--
    }
    let i1=0,j1=k-1
    while(i1<j1){
        // [nums[i1],nums[j1]]=[nums[j1],nums[i1]]
        let t=nums[i1]
        nums[i1]=nums[j1]
        nums[j1]=t
        i1++,j1--
    }
    let i2=k,j2=nums.length-1
    while(i2<j2){
        // [nums[i2],nums[j2]]=[nums[j2],nums[i2]]
        let t=nums[i2]
        nums[i2]=nums[j2]
        nums[j2]=t
        i2++,j2--
    }
    return nums
}

File: to-python/arrays/t.py
# move zeroes
def moveZeroes(nums):
    i=0
    for j in range(len(nums)):
        if(nums[j]!=0):
            t = nums[i]
            nums[i] = nums[j]
            nums[j] = t
            i+=1
    return nums


# print(moveZeroes([0,1,0,3,12]))
# print(moveZeroes([2,1]))

# majority element
def majorityElement(nums):
    return sorted(nums)[len(nums)//2]
# print(majorityElement([3,2,3]))
# an observation the .sort() method works on original array and modifies it and returns None
# where the sorted() returns new modified array

# remove duplicates from sorted array
def removeDuplicatesFromSortedArray(nums):
    # return len(set(nums))
    # works but fails for in place ones
    i=1
    for j in range(1,len(nums)):
        if(nums[j] != nums[j-1]):
            nums[i]=nums[j]
            i+=1
    return i
# print(removeDuplicatesFromSortedArray([1,1,2]))
# print(removeDuplicatesFromSortedArray([0,0,1,1,1,2,2,3,3,4]))

# best time to buy and sell stock
def bestTimeToBuyAnsSellStock(prices):
    # buy = prices[0]
    # ans=0
    # for j in range(1,len(prices)):
    #     buy = min(buy,prices[j])
    #     ans = max(ans,prices[j]-buy)
    # return ans
    # another better one
    i=0
    j=1
    ans=0
    while j< len(prices):
        if(prices[i]<prices[j]):
            ans = max(ans,prices[j]-prices[i])
        else:
            i=j
        j+=1
    return ans

# print(bestTimeToBuyAnsSellStock([7,1,5,3,6,4]))

# rotate array
def rotateArray(nums,k):
    k=k%len(nums)
    i=0
    j=len(nums)-1
    while(i<j):
        t=nums[i]
        nums[i]=nums[j]
        nums[j]=t
        i+=1
        j-=1
    i=0
    j=k-1
    while(i<j):
        t=nums[i]
        nums[i]=nums[j]
        nums[j]=t
        i+=1
        j-=1
    i=k
    j=len(nums)-1
    while(i<j):
        t=nums[i]
        nums[i]=nums[j]
        nums[j]=t
        i+=1
        j-=1
    return nums
# print(rotateArray([1,2,3,4,5,6,7],3))
# print(rotateArray([-1],2))

def product_of_array_except_itself(nums):
    length = len(nums)
    prefix=[0]*length
    suffix=[0]*length
    output=[0]*length
    prefix[0]=1
    for i in range(1,length):
        prefix[i]=nums[i-1]*prefix[i-1]
    suffix[length-1]=1
    for i in range(length-2,-1,-1):
        suffix[i]=nums[i+1]*suffix[i+1]
    for i in range(length):
        output[i]=prefix[i]*suffix[i]
    return output
    
print(product_of_array_except_itself([1,2,3,4]))